securetar/__init__.py

```diff
@@ -138,6 +168,94 @@ def size(self) -> float:
         return round(self._name.stat().st_size / 1_048_576, 2)  # calc mbyte
 
 
+class _InnerSecureTarFile(SecureTarFile):
+    """Handle encrypted files for tarfile library inside another tarfile."""
+
+    def __init__(
+        self,
+        outer_tar: tarfile.TarFile,
+        name: Path,
+        mode: str,
+        key: Optional[bytes] = None,
+        gzip: bool = True,
+        bufsize: int = DEFAULT_BUFSIZE,
+    ) -> None:
+        """Initialize inner handler."""
+        super().__init__(
+            name=name,
+            mode=mode,
+            key=key,
+            gzip=gzip,
+            bufsize=bufsize,
+            fileobj=outer_tar.fileobj,
+        )
+        self.outer_tar = outer_tar
+        self.stream: Generator[BinaryIO, None, None] | None = None
+
+    def __enter__(self) -> tarfile.TarFile:
+        """Start context manager tarfile."""
+        tar_info = tarfile.TarInfo(name=str(self._name))
+        tar_info.mtime = time.time()
+        self.stream = add_stream(self.outer_tar, tar_info)
+        self.stream.__enter__()
+        return super().__enter__()
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        """Close file."""
+        super().__exit__(exc_type, exc_value, traceback)
+        self.stream.__exit__(exc_type, exc_value, traceback)
+
+
+@contextmanager
+def add_stream(
+    tar: tarfile.TarFile, tar_info: tarfile.TarInfo
+) -> Generator[BinaryIO, None, None]:
+    """Add a stream to the tarfile.
+
+    This only works with uncompressed, unencrypted tar files.
+
+    The typical usage is:
+
+    with add_stream(tar, tar_info) as fileobj:
+        fileobj.write(data)
```

Comment:
Maybe add a warning not to mutate the Tarinfo in the context manager because it could corrupt internal state.  That might be overkill though as I don't think someone would do that ... but they will

Reply:
The only case where this seems to be an issue is when something sets the mtime after.

I was trying to avoid having to construct it again since it's not so cheap to build the header. For that reason I made the function private as the usage we have safe and has full test coverage since it is only called from the inner object.

Reply:
For context:

I made the api protected after I added that comment because I was worried someone would call stream_add directly and get it wrong.

All publicly exposed API are safe since we don't let the consumer of the api access the TarInfo object so there no opportunity to get it wrong

Reply:
I see yeah make sense. :+1: